### PR TITLE
RPC, stdcall and fastcall worked between Linux driver and OP-TEE OS

### PR DESCRIPTION
**Pull request by xyzhao2018**

### CODE REVIEW

**Comment by GangSecurity** [comment]
need rebase

File: core/arch/x86_64/mm/core_mmu.c [hunk 1]
```diff
@@ -7,7 +7,7 @@
 //#include <kernel/trace_control_by_service.h>
 #ifndef TRACE_SERV_MMU
 #undef TRACE_LEVEL
-#define TRACE_LEVEL 0
+#define TRACE_LEVEL 2
```
Comment (jingdlu) [inline core/arch/x86_64/mm/core_mmu.c]: Why do you comment out including trace_control_by_service.h and change TRACE_LEVEL to 2?

File: core/arch/x86_64/mm/core_mmu.c [hunk 2]
```diff
@@ -821,6 +821,11 @@ static size_t collect_mem_ranges(struct tee_mmap_region *memory_map,
 		/* Only unmapped virtual range may have a null phys addr */
 		assert(m.addr || !core_mmu_type_to_attr(m.type));
 
+		if (m.type == MEM_AREA_IO_NSEC || m.type == MEM_AREA_IO_SEC) {
+			m.addr = ROUNDDOWN(m.addr, CORE_MMU_PGDIR_SIZE);
+			m.size = ROUNDUP(m.size + (mem->addr - m.addr),CORE_MMU_PGDIR_SIZE);
+		}
```
Comment (jingdlu) [inline core/arch/x86_64/mm/core_mmu.c]: Why do you need special handling for MEM_AREA_IO_NSEC  in collect_mem_ranges function?  Does Arm have the same code?

File: core/arch/x86_64/mm/core_mmu.c [hunk 3]
```diff
@@ -894,8 +899,6 @@ static bool assign_mem_va(vaddr_t tee_ram_va,
 			map->va = va;
 			if (ADD_OVERFLOW(va, map->size, &va))
 				return false;
-			if (va >= BIT64(get_va_width()))
```
Comment (jingdlu) [inline core/arch/x86_64/mm/core_mmu.c]: Why do we need to remove this?

File: core/include/scattered_array.h [hunk 4]
```diff
@@ -83,7 +83,7 @@
  * @element_type: The type of the elemenet
  */
 #define SCATTERED_ARRAY_BEGIN(array_name, element_type) (__extension__({ \
-		static const element_type __scattered_array_begin[0] __unused \
+		static const element_type __scattered_array_begin[0] __aligned(32) __unused \
```
Comment (jingdlu) [inline core/include/scattered_array.h]: This is a common file. Do we need to change this file?

File: core/arch/x86_64/kernel/thread_optee_smc.c [hunk 5]
```diff
@@ -300,17 +300,17 @@ static struct mobj *thread_rpc_alloc_arg(size_t size)
 {
 	paddr_t pa;
 	uint64_t co;
-	uint32_t rpc_args[THREAD_RPC_NUM_ARGS] = {
-		OPTEE_SMC_RETURN_RPC_ALLOC, size
+	uint32_t rpc_args[THREAD_RPC_NUM_ARGS]= {
```
Comment (GangSecurity) [inline core/arch/x86_64/kernel/thread_optee_smc.c]: Add original space here

**Review by GangSecurity — COMMENTED** [review]
Vmcall from TEE should send to core 0 in REE. This patch miss the code

**Comment by xyzhao2018** [comment]
> Vmcall from TEE should send to core 0 in REE. This patch miss the code

It is optee Linux driver in REE which must send vmcall on core 0 to TEE. TEE always run on core 0.

**Review by GangSecurity — APPROVED** [review]

File: core/arch/x86_64/kernel/thread_private.h [hunk 6]
```diff
@@ -169,7 +169,7 @@ void thread_unlock_global(void);
  * The purpose of this function is to request services from non-secure
  * world.
  */
-#define THREAD_RPC_NUM_ARGS     4
+#define THREAD_RPC_NUM_ARGS     6
```
Comment (jingdlu) [inline core/arch/x86_64/kernel/thread_private.h:172]: Why do we still need 6 args number here?

File: core/arch/x86_64/kernel/thread_optee_smc.c [hunk 7]
```diff
@@ -301,16 +301,16 @@ static struct mobj *thread_rpc_alloc_arg(size_t size)
 	paddr_t pa;
 	uint64_t co;
 	uint32_t rpc_args[THREAD_RPC_NUM_ARGS] = {
-		OPTEE_SMC_RETURN_RPC_ALLOC, size
+		OPTEE_SMC_RETURN_RPC_ALLOC, (uint32_t)(size & 0x00000000FFFFFFFF)
```
Comment (jingdlu) [inline core/arch/x86_64/kernel/thread_optee_smc.c:304]: This is uint32_t array. Do we need this type casting?

Comment (xyzhao2018) [reply core/arch/x86_64/kernel/thread_optee_smc.c:304]: Yes, otherwise, the data is not correct.

Comment (jingdlu) [reply core/arch/x86_64/kernel/thread_optee_smc.c:304]: Do you know why there is no this issue on Arm or OP-TEE 3.4 x86?

Comment (xyzhao2018) [reply core/arch/x86_64/kernel/thread_optee_smc.c:304]: Currently, don't know, we make it work firstly. 
Will do some research later, maybe related to compile process.

File: core/arch/x86_64/kernel/thread_private.h (hunk 6 above)
Comment (xyzhao2018) [reply core/arch/x86_64/kernel/thread_private.h:172]: ARM doesn't use parameter a4, a5, just cut them. 

	/*
	 * Return from RPC to request service of a foreign interrupt must not
	 * get parameters from non-secure world.
	 */
	if (threads[n].flags & THREAD_FLAGS_COPY_ARGS_ON_RETURN) {
		copy_a0_to_a3(&threads[n].regs, a0, a1, a2, a3);
		threads[n].flags &= ~THREAD_FLAGS_COPY_ARGS_ON_RETURN;
	}

**Review by jingdlu — APPROVED** [review]

**Pull request merged**